File: Algorithms/Greedy/Huffman.java

import java.util.*;

public class Huffman {

    private static class Node implements Comparable<Node> {

        Node left, right;
        char value;
        int freq;

        public Node(char ch, int f) {

            value = ch;
            freq = f;

        }

        @Override
        public int compareTo(Node other) {

            return freq - other.freq;

        }

    }

    private static Map<Character, Integer> getFreqMap(String msg) {

        Map<Character, Integer> map = new TreeMap<>();

        for (char c : msg.toCharArray()) {

            if (map.containsKey(c)) {

                map.put(c, map.get(c) + 1);

            } else {

                map.put(c, 1);

            }

        }

        return map;

    }

    private static void addCodes(Node node, Map<Character, String> codes, String prefix) {

        if (node != null) {

            if (node.value != '*') {

                codes.put(node.value, prefix);

            }

            addCodes(node.left, codes, prefix + "0");
            addCodes(node.right, codes, prefix + "1");

        }

    }

    public static Map<Character, String> getHuffmanCodes(String msg) {

        Map<Character, Integer> freqs = getFreqMap(msg);

        PriorityQueue<Node> heap = new PriorityQueue<>();

        for (char c : freqs.keySet()) {

            heap.add(new Node(c, freqs.get(c)));

        }

        while (heap.size() > 1) {

            Node left = heap.remove(), right = heap.remove();

            Node combined = new Node('*', left.freq + right.freq);

            combined.left = left;
            combined.right = right;

            heap.add(combined);

        }

        Map<Character, String> codes = new TreeMap<>();

        addCodes(heap.peek(), codes, "");

        return codes;

    }

    public static String encode(String msg, Map<Character, String> codes) {

        String compressed = "";

        for (char c : msg.toCharArray()) {

            compressed += codes.get(c);

        }

        return compressed;

    }

    public static void main(String[] args) {

        String msg = "This is a test of huffman encoding............";

        System.out.println(msg);

        Map<Character, String> codes = getHuffmanCodes(msg);

        for (char c : codes.keySet()) {

            System.out.printf("%s %s%n", c, codes.get(c));

        }

        System.out.println(encode(msg, codes));

    }

}

File: Algorithms/Greedy/Huffman.js
"use strict";

class Heap { // Insert Actual Heap Here...

    constructor() {

        this.items = [];
        this.nodeSort = (a, b) => a.freq - b.freq;

    }

    push(item) {

        this.items.push(item);
        this.items.sort(this.nodeSort);

    }

    remove() {

        return this.items.shift();

    }

    get size() {

        return this.items.length;

    }

}

class Node {

    constructor(ch, f) {

        this.value = ch;
        this.freq = f;
        this.left = null;
        this.right = null;

    }

}

let getFreqMap = (msg) => {

    let map = {};

    for (let c of msg) {

        if (c in map) {

            map[c]++;

        } else {

            map[c] = 1;

        }

    }

    return map;

}

let addCodes = (node, codes, prefix) => {

    if (node != null) {

        if (node.value != '*') {

            codes[node.value] = prefix;

        }

        addCodes(node.left, codes, prefix + "0");
        addCodes(node.right, codes, prefix + "1");

    }

}

let getHuffmanCodes = (msg) => {

    let freqs = getFreqMap(msg);

    let heap = new Heap();

    for (let c in freqs) {

        heap.push(new Node(c, freqs[c]));

    }

    while (heap.size > 1) {

        let left = heap.remove(),
            right = heap.remove();

        let combined = new Node('*', left.freq + right.freq);

        combined.left = left;
        combined.right = right;

        heap.push(combined);

    }

    let codes = {};

    addCodes(heap.remove(), codes, "");

    return codes;

}

let encode = (msg, codes) => {

    let compressed = "";

    for (let c of msg) {

        compressed += codes[c];

    }

    return compressed;

}

let msg = "This is a test of huffman encoding............";

console.log(msg);

let codes = getHuffmanCodes(msg);

for (let c in codes) {

    console.log(`${c} ${codes[c]}`);

}

console.log(encode(msg, codes));
